### DISCUSSION TITLE
Where is data stored for offline users? (Android)

### DISCUSSION THREAD

**Discussion by logansc** [standardnotes/forum#2751] [General]
I made a backup of my notes and uninstalled the app. Unfortunately the backup that StandardNotes created is partially corrupted, so I want to try and recover the files from my phone. I'm willing to root the device but I need to know where the data was stored in order for this to (possibly) work.  I'm using Android on a Oneplus 6.  Thank you.

**Comment by atmoio** [comment]
I have this saved in my notes, but it's from 2019. Hopefully it can serve as a good starting point. Can you elaborate how the backup file you have is corrupted?

1. Run `adb backup -noapk com.standardnotes` and confirm on device. A file named `backup.ab` will be created in your current directory. Move the file to desktop for convenience.

2. Download android-backup-extractor from here: https://sourceforge.net/projects/adbextractor/

3. Navigate to `android-backup-tookit/android-backup-extractor/android-backup-extractor-20180521-bin` and find `abe.jar`. 

4. Move abe.jar to your desktop.

5. Run `java -jar abe.jar unpack backup.ab backup.tar`

6. backup.tar will now exist on your desktop. Unarchive that.

7. Navigate to apps/com.standardnotes/db/RKStorage and rename file to RKStorage.sqlite.

8. Open the sqlite file with an application that can handle it.

Resources: 
- https://stackoverflow.com/questions/13006315/how-to-access-data-data-folder-in-android-device
- https://forum.xda-developers.com/showthread.php?t=2011811

**Comment by logansc** [comment]
Thanks, I really appreciate the response.  Unfortunately when I followed your
instructions, the resulting backup.ab file was 0 bytes; I'm assuming that
fstrim must have run before I turned the device off or when I started it.
As far as the corrupted backup, it looks like the corruption starts where a
question mark character was.  Then at another "?" character, the corruption
ends.  It does this several times in the notes, always starting and ending
where question marks were.  Strangely, there are also question marks in the
document that don't cause this.  Here's a sample:

What if people think it's a bug — — ’ Û H ˜\Ú W — — ”Ý Ü ] W — — • H ™YH ^
\È \ Ûˆ Ú XÚ [Ý\ˆ ™YKY [Y[œÚ[Û˜[ X] š^ Ùˆ ÛÛZXÜÈ Ú[ ™H Z[

Also, when I take the bits of the corrupted characters, delete the first
two, then convert them back to ASCII, some of the data becomes partially
coherent.  Here is a sample of the above text when I do that:

\o� Po� P#� do ly ts.qs h!\#� Po� P#� w to ppit!\#� Po� Po� �he ths� �e
axes upo o. w hic h yo us. ths� �e-dimeo O io o.ql mat9

I recognize some of the snippets of words from my old notes in there.  I
also noticed that many of the characters have been incremented to the next
ASCII code (n -> o, r -> s, etc.).  It's not much to go on, but if you have
any insight I'd love to hear it.  If not, I appreciate the help you've
already given and wish you the best.

**Comment by atmoio** [comment]
Hmm yeah I'm afraid I don't know what else it could be :/ Hard to tell if the original file is corrupted or if there is some encoding issue.
